pd/src/app.rs

```diff
@@ -443,24 +452,63 @@ impl App {
 
                 let mut next_rates = Vec::new();
                 let mut next_validator_statuses = Vec::new();
+
+                // this is a bit complicated: because we're in the EndBlock phase, and the
+                // delegations in this block have not yet been committed, we have to combine
+                // the delegations in pending_block with the ones already committed to the
+                // state. otherwise the delegations committed in the epoch threshold block
+                // would be lost.
+                let mut committed_delegation_changes =
```

Comment:
Should this maybe be called `delegation_changes` rather than `committed_delegation_changes`, if we're adding in the delegation changes from the pending block that haven't been committed yet?